epfmt.1.md
---
title: "epfmt (1) user manual"
author: "R. S. Doiel"
pubDate: 2023-03-03
---

# NAME

epfmt

# SYNOPSIS

epfmt 

# DESCRIPTION

epfmt is a command line program for pretty printing
EPrint XML. It can also convert EPrint XML to and from
EPrint as JSON or as a simplified JSON records for export
to RDM. 

By default it reads from standard input and writes to
standard out.

epfmt EPrint XML (or JSON version) from
standard input and pretty prints the result to 
standard out. You can change output format XML 
and JSON by using either the '-xml' or '-json' 
option. The XML representation is based on EPrints 
3.x.  epfmt does NOT interact with the EPrints API 
only the the document presented via standard
input.

# OPTIONS

-help
: display help

-license
: display license

-i, -input
: (string) input file name (read the URL connection string from the input file

-json
: output JSON version of EPrint XML

-nl, -newline
: if true add a trailing newline

-o, -output
: (string) output file name

-quiet
: suppress error messages

-s, -simple
: output simplified JSON version of EPrints XML

-version
: display version

-xml
: output EPrint XML

# EXAMPLES

Pretty print EPrint XML as XML.

~~~
    epfmt < 123.xml
~~~

Pretty print from EPrint XML as JSON

~~~
    epfmt -json < 123.xml
~~~

Render EPrint JSON as EPrint XML.

~~~
    epfmt -xml < 123.json
~~~

epfmt will first parse the XML or JSON 
presented to it and pretty print the output 
in the desired format requested. If no 
format option chosen it will pretty print 
in the same format as input.

epfmt 1.2.4




CHANGES.md
Release 1.1.0:

Significant revisition of the internal data structures in the
EPrinttools package.  Improvements in doi2eprintxml file handling,
ruleset are implemented with individual rule selection possible.
A implementation of ep3apid (an EPrints 3.3.16 extended API) includes
support for importing metadata into EPrints repositories. An increase
in the end points supported.

Release 1.0.3-next:

Added `/<REPO_ID>/keys` to list all keys in a repository directly from the MySQL tables
Added `/<REPO_ID>/eprint/<EPRINT_ID` (GET) for generating EPrint XML directly form the MySQL tables

The eprint3x.go has been heavely updated. Included are corrections, deletions and additions in the various struct definitions. Added an `ItemsInterface` type so I can consolidate the SQL handling around common item list types.

The ep3sql.go file contains all the SQL interaction mapping relational models to our structs.

The ep3apid now only uses MySQL connection for returning results (not the EPrints REST API).  

Release 1.0.2-simplified:

Added `/<REPO_ID>/created/<TIMESTAMP>/<TIMESTAMP>` end point based on
eprint table's "datestamp" fields.

Release 1.0.2:

Introduced a web service called `ep3apid` which can be run from the
command line or setup to run from systemd (Debian Linux) or 
launchd (macOS). It provides a local host web service for quick
key list retrieval by talking directly to the EPrints MySQL database(s).

Introduced a "simplified" JSON model based on DataCite and Invenio 3.
This is supported in `eputil`, `epfmt` using command line options
and is the default JSON record output for `ep3apid`.

Depreciated `eprintxml2json`, superceded by `epfmt`.

By default clsrules now require the command line option to be applied.

Removed Python experiments from this repository

Removed dependencies on "github.com/caltechlibrary/cli" and
"github.com/caltechlibrary/rc".

Release 1.0.1:

- Depreciating the `-clsrules=true` option. when using `doi2eprintxml`. Before release 1.0.1 the DOI was included by default in the related url field. This was an artifact of Caltech Library support DOI before EPrints did. This practice ended in 2021. Additional haviors like limiting the number of authors ended in 2020.  To reflect these changes in practive the `-clsrules` option now defaults to "false" and these rules are not applied unless you explicitly add `-clsrules` on the command line.
    - See clsrules/README.md for details
- Added in release 1.0.1 is an option to map ISSN to publisher names overriding what is provided by CrossRef and DataCite. To include include this option choose `-issn-to-publisher` option.  This option will do two things, it will create an `issn-to-publisher.csv` if none existed and if one does exist it will read it in using that mapping. In this way we can control how publishers are named.


cmd/epfmt/epfmt.go
// Package eprinttools is a collection of structures, functions and programs// for working with the EPrints XML and EPrints REST API
//
// @author R. S. Doiel, <rsdoiel@caltech.edu>
//
// Copyright (c) 2021, Caltech
// All rights not granted herein are expressly reserved by Caltech.
//
// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:
//
// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
//
// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
//
// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
package main

//
// epfmt is a command line tool convert EPrints XML to/from JSON.
//

import (
	"bytes"
	"encoding/json"
	"encoding/xml"
	"flag"
	"fmt"
	"io/ioutil"
	"os"
	"path"
	"strings"

	// Caltech Library Packages
	"github.com/caltechlibrary/eprinttools"
)

const (
	IsXML = iota
	IsJSON
)

var (
	helpText = `---
title: "{app_name} (1) user manual"
author: "R. S. Doiel"
pubDate: 2023-03-03
---

# NAME

{app_name}

# SYNOPSIS

{app_name} 

# DESCRIPTION

{app_name} is a command line program for pretty printing
EPrint XML. It can also convert EPrint XML to and from
EPrint as JSON or as a simplified JSON records for export
to RDM. 

By default it reads from standard input and writes to
standard out.

{app_name} EPrint XML (or JSON version) from
standard input and pretty prints the result to 
standard out. You can change output format XML 
and JSON by using either the '-xml' or '-json' 
option. The XML representation is based on EPrints 
3.x.  {app_name} does NOT interact with the EPrints API 
only the the document presented via standard
input.

# OPTIONS

-help
: display help

-license
: display license

-i, -input
: (string) input file name (read the URL connection string from the input file

-json
: output JSON version of EPrint XML

-nl, -newline
: if true add a trailing newline

-o, -output
: (string) output file name

-quiet
: suppress error messages

-s, -simple
: output simplified JSON version of EPrints XML

-version
: display version

-xml
: output EPrint XML

# EXAMPLES

Pretty print EPrint XML as XML.

~~~
    {app_name} < 123.xml
~~~

Pretty print from EPrint XML as JSON

~~~
    {app_name} -json < 123.xml
~~~

Render EPrint JSON as EPrint XML.

~~~
    {app_name} -xml < 123.json
~~~

{app_name} will first parse the XML or JSON 
presented to it and pretty print the output 
in the desired format requested. If no 
format option chosen it will pretty print 
in the same format as input.

{app_name} {version}

`

	// Standard Options
	showHelp         bool
	showLicense      bool
	showVersion      bool
	showExamples     bool
	newLine          bool
	quiet            bool
	verbose          bool
	generateMarkdown bool
	generateManPage  bool
	inputFName       string
	outputFName      string

	// App Options
	asJSON       bool
	asXML        bool
	asSimplified bool
)

func fmtTxt(src string, appName string, version string) string {
	return strings.ReplaceAll(strings.ReplaceAll(src, "{app_name}", appName), "{version}", version)
}

func main() {
	var (
		inputFmt int
		obj      *eprinttools.EPrints
		src      []byte
	)

	appName := path.Base(os.Args[0])

	// Standard Options
	flag.BoolVar(&showHelp, "help", false, "display help")
	flag.BoolVar(&showLicense, "license", false, "display license")
	flag.BoolVar(&showVersion, "version", false, "display version")
	flag.StringVar(&inputFName, "i", "", "input file name (read the URL connection string from the input file")
	flag.StringVar(&inputFName, "input", "", "input file name (read the URL connection string from the input file")
	flag.StringVar(&outputFName, "o", "", "output file name")
	flag.StringVar(&outputFName, "output", "", "output file name")
	flag.BoolVar(&quiet, "quiet", false, "suppress error messages")
	flag.BoolVar(&newLine, "nl", false, "if true add a trailing newline")
	flag.BoolVar(&newLine, "newline", false, "if true add a trailing newline")

	// App Options
	flag.BoolVar(&asXML, "xml", false, "output EPrint XML")
	flag.BoolVar(&asJSON, "json", false, "output JSON version of EPrint XML")
	flag.BoolVar(&asSimplified, "s", false, "output simple JSON record version of EPrints XML")
	flag.BoolVar(&asSimplified, "simple", false, "output simple JSON record version of EPrints XML")

	// We're ready to process args
	flag.Parse()
	args := flag.Args()

	// Simplified output is aways JSON formatted.
	if asSimplified {
		asJSON = true
		asXML = false
	}

	// Setup IO
	var err error

	in := os.Stdin
	out := os.Stdout
	eout := os.Stderr

	if len(args) > 1 {
		inputFName = args[1]
	}
	if len(args) > 2 {
		outputFName = args[2]
	}

	if inputFName != "" && inputFName != "-" {
		in, err = os.Open(inputFName)
		if err != nil {
			fmt.Fprintln(eout, err)
			os.Exit(1)
		}
		defer in.Close()
	}

	if outputFName != "" && outputFName != "-" {
		out, err = os.Create(outputFName)
		if err != nil {
			fmt.Fprintln(eout, err)
			os.Exit(1)
		}
		defer out.Close()
	}

	// Handle options
	if showHelp {
		fmt.Fprintf(out, "%s\n", fmtTxt(helpText, appName, eprinttools.Version))
		os.Exit(0)
	}
	if showLicense {
		fmt.Fprintf(out, "%s\n", eprinttools.LicenseText)
		os.Exit(0)
	}
	if showVersion {
		fmt.Fprintf(out, "%s %s\n", appName, eprinttools.Version)
		os.Exit(0)
	}

	// Read the file to []byte
	src, err = ioutil.ReadAll(in)
	if err != nil {
		fmt.Fprintln(eout, err)
		os.Exit(1)
	}
	// Trim leading/trailing spaces
	src = bytes.TrimSpace(src)
	if len(src) == 0 {
		fmt.Fprintln(eout, "Nothing to parse")
		os.Exit(1)
	}

	// Check if JSON or XML
	if bytes.HasPrefix(src, []byte("{")) {
		// Unmarshal as JSON
		inputFmt = IsJSON
		err = json.Unmarshal(src, &obj)
	} else {
		// Unmarshal as EPrintXML
		inputFmt = IsXML
		err = xml.Unmarshal(src, &obj)
	}
	if err != nil {
		fmt.Fprintln(eout, err)
		os.Exit(1)
	}

	for _, e := range obj.EPrint {
		e.SyntheticFields()
	}
	if asJSON == false && asXML == false {
		asXML = (inputFmt == IsXML)
	}
	if asSimplified {
		if len(obj.EPrint) == 1 {
			sObject := new(eprinttools.Record)
			err := eprinttools.CrosswalkEPrintToRecord(obj.EPrint[0], sObject)
			if err != nil {
				fmt.Fprintln(eout, err)
				os.Exit(1)
			}
			src, err = json.MarshalIndent(sObject, "", "   ")
		} else {
			sObjects := []*eprinttools.Record{}
			for _, eprint := range obj.EPrint {
				obj := new(eprinttools.Record)
				err := eprinttools.CrosswalkEPrintToRecord(eprint, obj)
				if err != nil {
					fmt.Fprintln(eout, err)
					os.Exit(1)
				}
				sObjects = append(sObjects, obj)
			}
			src, err = json.MarshalIndent(sObjects, "", "   ")
		}
	} else {
		// marshal pretty printed output based on options selected.
		if asXML {
			src, err = xml.MarshalIndent(obj, "", "  ")
		} else {
			src, err = json.MarshalIndent(obj, "", "   ")
		}
	}
	if err != nil {
		fmt.Fprintln(eout, err)
		os.Exit(1)
	}
	// Print the doc string if XML
	if asXML {
		fmt.Fprintf(out, "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n")
	}

	eol := ""
	if newLine {
		eol = "\n"
	}
	fmt.Fprintf(out, "%s%s", src, eol)
	os.Exit(0)
}
